feat(browser): Add browser metrics sdk

AbhiPrasad opened the pull request:
Please read the following before reviewing:
- https://develop.sentry.dev/sdk/metrics/
- https://develop.sentry.dev/delightful-developer-metrics/

ref https://github.com/getsentry/sentry-javascript/issues/9691

This PR introduces functionality for a browser metrics SDK in an alpha state. Via the newly introduced APIs, you can now flush metrics directly to Sentry in the browser.

To enable capturing metrics, you first need to add the `Metrics` integration. This is done for treeshaking purposes.

```js
Sentry.init({
  dsn: '__DSN__',
  integrations: [
    new Sentry.metrics.MetricsAggregator(),
  ],
});
```

Then you'll be able to add `counters`, `sets`, `distributions`, and `gauges` under the `Sentry.metrics` namespace.

```js
// Add 4 to a counter named `hits`
Sentry.metrics.increment('hits', 4);

// Add 2 to gauge named `parallel_requests`, tagged with `happy: "no"`
Sentry.metrics.gauge('parallel_requests', 2, { tags: { happy: 'no' } });

// Add 4.6 to a distribution named `response_time` with unit seconds
Sentry.metrics.distribution('response_time', 4.6, { unit: 'seconds' });

// Add 2 to a set named `valuable.ids`
Sentry.metrics.set('valuable.ids', 2);
```

Under the hood, adding the `Metrics` integration adds a `SimpleMetricsAggregator`. This is a bundle-sized efficient metrics aggregator that aggregates metrics and flushes them out every 5 seconds. This does not use any weight based logic - this will be implemented in the `MetricsAggregator` that is used for server runtimes (node, deno, vercel-edge).

To make metrics conditional, it is defined on the client as `public metricsAggregator: MetricsAggregator | undefined`.

Next step is to add some integration tests for this functionality, and then add server-runtime logic for it.

Many of the decisions taken for this aggregator + flushing implementation was to try to be as bundle size efficient as possible, happy to answer any specific questions.

anonrig commented on packages/core/src/baseclient.ts:
```diff
@@ -278,6 +290,9 @@ export abstract class BaseClient<O extends ClientOptions> implements Client<O> {
   public close(timeout?: number): PromiseLike<boolean> {
     return this.flush(timeout).then(result => {
       this.getOptions().enabled = false;
+      if (this.metricsAggregator) {
+        this.metricsAggregator.close();
+      }
```
```suggestion
      this.metricsAggregator?.close();
```

anonrig replied:
We don't use optional chaining don't we? I always forget!

anonrig commented on packages/core/src/metrics/integration.ts:
```diff
@@ -0,0 +1,38 @@
+  public name: string;
+
+  public constructor() {
+    this.name = Metrics.id;
```
Can we move this to line 19?

anonrig commented on packages/core/src/metrics/simpleaggregator.ts:
```diff
@@ -0,0 +1,120 @@
+  private readonly _interval: ReturnType<typeof setInterval>;
+
+  public constructor(private readonly _client: Client<ClientOptions>) {
+    this._buckets = new Map();
```
Can you move this to line 28?

anonrig commented on packages/core/src/metrics/simpleaggregator.ts:
```diff
@@ -0,0 +1,120 @@
+    unsanitizedName: string,
+    value: number,
+    unit: MeasurementUnit = 'none',
+    unsanitizedTags: { [key: string]: Primitive } = {},
```
```suggestion
    unsanitizedTags: Record<string, Primitive> = {},
```

anonrig commented on packages/core/src/metrics/simpleaggregator.ts:
```diff
@@ -0,0 +1,120 @@
+ */
+export function serializeBuckets(buckets: SimpleMetricBucket): string {
+  let out = '';
+  buckets.forEach(([metric, timestamp, metricType, name, unit, tags]) => {
```
Nit: We should avoid using forEach. `for (let...)` is always faster

anonrig commented on packages/core/src/metrics/simpleaggregator.ts:
```diff
@@ -0,0 +1,120 @@
+  return out;
+}
+
+function sanitizeTags(unsanitizedTags: { [key: string]: Primitive }): { [key: string]: string } {
```
```suggestion
function sanitizeTags(unsanitizedTags: Record<string, Primitive>): { [key: string]: string } {
```

anonrig commented on packages/core/src/metrics/simpleaggregator.ts:
```diff
@@ -0,0 +1,120 @@
+function sanitizeTags(unsanitizedTags: { [key: string]: Primitive }): { [key: string]: string } {
+  const tags: { [key: string]: string } = {};
+  for (const key in unsanitizedTags) {
+    if (Object.prototype.hasOwnProperty.call(unsanitizedTags, key)) {
```
Any reason why not the following?

```suggestion
    if (unsanitizedTags.hasOwn(key)) {
```

AbhiPrasad replied:
Doesn't work for ie11 😢

anonrig commented on packages/core/src/metrics/types.ts:
```diff
@@ -0,0 +1,36 @@
+    name: string,
+    value: number,
+    unit?: MeasurementUnit,
+    tags?: { [key: string]: Primitive },
```
```suggestion
    tags?: Record<string, Primitive>,
```

lforst commented on packages/types/src/client.ts:
```diff
@@ -179,6 +179,13 @@ export interface Client<O extends ClientOptions = ClientOptions> {
+   *
+   * @experimental This API is experimental and might experience breaking changes
+   */
+  captureSerializedMetrics?(serializedMetrics: string): void;
```
l/m: It honestly feels kinda weird that we pass serialized data in this client method. For all the other event types we pass unserialized data 🤔 Should we rethink this?

AbhiPrasad replied:
Couldn't think of a better way to do this without adding to bundle size. I'm all ears if you have an alternative.

lforst replied:
I don't know how it would affect bundle size but we could maybe turn this into `captureMetrics(metrics: AggregatedMetrics)`. This would just be to be a bit more "conventional" though. User benefit is probably slim to none. Maybe this would tie a bit into my point about having an Aggregator only do aggregation and no creation of metrics...

lforst commented on packages/core/src/metrics/instance.ts:
```diff
@@ -0,0 +1,106 @@
+
+  /** JSDoc */
+  public toString(): string {
+    return `${Array.from(this.value).join(':')}`;
```
Okay apparently we need to at least convert to 32bit or 64bit integers.

AbhiPrasad commented on packages/core/src/metrics/simpleaggregator.ts:
```diff
@@ -0,0 +1,120 @@
+ * Serialize metrics buckets into a string based on statsd format.
+ */
+export function serializeBuckets(buckets: SimpleMetricBucket): string {
+  let out = '';
```
done - b05729edeb2b72276dc161707134ef6ef61e0248

AbhiPrasad commented on packages/core/src/metrics/simpleaggregator.ts:
```diff
@@ -0,0 +1,120 @@
+ *
+ * @experimental This API is experimental and might change in the future.
+ */
+export class SimpleMetricsAggregator implements MetricsAggregator {
```
There is also possibility to use alternative metrics aggregator for react native/electron

lforst commented on packages/core/src/metrics/simpleaggregator.ts:
```diff
@@ -0,0 +1,120 @@
+  /**
+   * @inheritDoc
+   */
+  public add(
```
Ok now that I thought about it a bit more. I think your point makes sense!

mydea commented on packages/core/src/baseclient.ts:
```diff
@@ -264,6 +273,9 @@ export abstract class BaseClient<O extends ClientOptions> implements Client<O> {
   public flush(timeout?: number): PromiseLike<boolean> {
     const transport = this._transport;
     if (transport) {
+      if (this.metricsAggregator) {
```
I wonder, would it maybe be nicer to solve this via hooks?

1. The metrics integration registers handlers for `beforeFlush`, `afterFlush`, `close` (or something like this)
2. The client emits these 

WDYT?

AbhiPrasad replied:
I avoided this because I felt the metrics aggregator encapsulated the internal state relatively well, and matched up with the other sdk implementations.

In addition, this is now super easy to make work once we swap in another metrics aggregator implementation for server runtime in https://github.com/getsentry/sentry-javascript/issues/9797

mydea commented on packages/core/src/metrics/exports.ts:
```diff
@@ -0,0 +1,81 @@
+ *
+ * @experimental This API is experimental and might having breaking changes in the future.
+ */
+export function set(name: string, incomingValue: number | string, data?: MetricData): void {
```
Maybe we can/should find a different name here? As `Sentry.set()` feels a bit weird to me 🤔

AbhiPrasad commented on packages/core/src/metrics/exports.ts:
```diff
@@ -0,0 +1,81 @@
+ *
+ * @experimental This API is experimental and might having breaking changes in the future.
+ */
+export function incr(name: string, value: number = 1, data?: MetricData): void {
```
This is up for debate - asking on slack about it.

AbhiPrasad replied:
Updated to increment.

AbhiPrasad commented on packages/browser/src/exports.ts:
```diff
@@ -56,6 +56,11 @@ export {
   withScope,
   FunctionToString,
   InboundFilters,
+  incr,
+  distribution,
+  set,
+  gauge,
+  Metrics,
```
Asked in slack, I think it's worthwhile to get better API naming here.

AbhiPrasad commented on packages/core/src/metrics/instance.ts:
> Would there be a size advantage when if we create PoJos with methods instead of the classes

Probably, but I elected to minimize bugs and match other SDK implementations more closely, so I chose classes (makes it easier to review since you can compare with other sdk). We can always go back and switch from class to POJO because this is completely hidden from users, open to be more optimized in the future.

We also care less about es5 nowadays so the es6 class is actually not too much larger than pojo, and classes have performance benefits in certain scenarios.

AbhiPrasad commented:
Updated the API:

```ts
// Add 4 to a counter named `hits`
Sentry.metrics.increment('hits', 4);

// Add 2 to gauge named `parallel_requests`, tagged with `happy: "no"`
Sentry.metrics.gauge('parallel_requests', 2, { tags: { happy: 'no' } });

// Add 4.6 to a distribution named `response_time` with unit seconds
Sentry.metrics.distribution('response_time', 4.6, { unit: 'seconds' });

// Add 2 to a set named `valuable.ids`
Sentry.metrics.set('valuable.ids', 2);
```

Also fixed so we don't override user set tags.

AbhiPrasad commented:
I went ahead and also refactored client and serialization logic as per @lforst feedback. It does have increased bundle size hit, but I believe it is waaaay more maintainable, thanks for keeping me honest!

https://github.com/getsentry/sentry-javascript/pull/9794#discussion_r1423689646

AbhiPrasad commented:
example envelope: 

```
{"sent_at":"2023-12-12T16:38:53.929Z","sdk":{"name":"sentry.javascript.browser","version":"7.86.0"}}
{"type":"statsd","length":260}
hits@none:4|c|#environment:production,transaction:/|T1702399133
parallel_requests@none:4:4:4:4:1|g|#environment:production,transaction:/,happy:no|T1702399133
response_time@none:0.2521277168899102|d|#environment:production,transaction:/,happy:maybe|T1702399133
```

In the UI: 

<img width="1265" alt="image" src="https://github.com/getsentry/sentry-javascript/assets/18689448/93f4781e-f02e-4a63-bfd6-cc6761636987">

AbhiPrasad commented on packages/core/src/metrics/envelope.ts:
```diff
@@ -27,17 +25,15 @@ export function createMetricEnvelope(
 function createMetricEnvelopeItem(metricAggregate: string): StatsdItem {
   const metricHeaders: StatsdItem[0] = {
     type: 'statsd',
+    content_type: 'application/octet-stream',
```
Removed, we don't need it.

Lms24 commented on packages/core/src/metrics/envelope.ts:
```diff
@@ -0,0 +1,40 @@
+    };
+  }
+
+  if (!!tunnel && !!dsn) {
```
l: do we need the double `!!` here? we just check for definedness, no?

AbhiPrasad replied:
we do not, will change.

cleptric commented on packages/core/src/metrics/constants.ts:
```diff
@@ -0,0 +1,30 @@
+export const NAME_AND_TAG_KEY_NORMALIZATION_REGEX = /[^a-zA-Z0-9_/.-]+/g;
+
+/**
+ * Normalization regex for metric tag balues.
```
```suggestion
 * Normalization regex for metric tag values.
```

mydea commented on packages/core/src/metrics/exports.ts:
```diff
@@ -0,0 +1,79 @@
+    const metricTags = {
+      ...tags,
+    };
+    if (release) {
+      metricTags.release = release;
+    }
+    if (environment) {
+      metricTags.environment = environment;
+    }
+    if (transaction) {
+      metricTags.transaction = transaction.name;
+    }
```
note that this is also somewhat like this for general events 🤔 e.g.

```js
// in scope applyToEvent
  if (this._level) {
      event.level = this._level;
    }
    if (this._transactionName) {
      event.transaction = this._transactionName;
    }
```

mydea replied:
... but only for some things, for others (e.g. tags etc.) event data takes precedence.

AbhiPrasad commented on packages/core/src/metrics/simpleaggregator.ts:
```diff
@@ -0,0 +1,120 @@
+
+  public constructor(private readonly _client: Client<ClientOptions>) {
+    this._buckets = new Map();
+    this._interval = setInterval(() => this.flush(), DEFAULT_FLUSH_INTERVAL);
```
Jitter is needed for server-side sdks when you have a fleet of machines all running the same sdk version. You want some random delay between the flush intervals of all those machines to reduce rate limiting/overloading Sentry (imagine 10 machines all sending to Sentry at the same time - multiply that by all the different uses of the sdk out there)

anonrig commented on packages/core/src/metrics/instance.ts:
```diff
@@ -0,0 +1,106 @@
+    this._min = value;
+    this._max = value;
+    this._sum = value;
+    this._count = 1;
```
This can be moved to line 30

AbhiPrasad replied:
For now it can't see https://github.com/getsentry/sentry-javascript/pull/8700

anonrig commented on packages/core/src/metrics/instance.ts:
```diff
@@ -0,0 +1,106 @@
+    this._min = Math.min(this._min, value);
+    this._max = Math.max(this._max, value);
+    this._sum += value;
+    this._count += 1;
```
Nit

```suggestion
    this._count++;
```

anonrig commented on packages/core/src/metrics/instance.ts:
```diff
@@ -0,0 +1,106 @@
+  public add(value: number): void {
+    this._last = value;
+    this._min = Math.min(this._min, value);
+    this._max = Math.max(this._max, value);
```
Instead of calling Math.min and later Math.max, a single if statement to understand which is smaller/bigger, would be a lot of faster.

anonrig commented on packages/browser/src/exports.ts:
```diff
@@ -56,6 +56,8 @@ export {
   FunctionToString,
   InboundFilters,
+  metrics,
+  Metrics,
```
Since this is a type why don't we include with import type statement?

AbhiPrasad replied:
`Metrics` is not a type, it is a class.

```ts
Sentry.init({
  dsn: '__DSN__',
  integrations: [
    new Sentry.Metrics(),
  ],
});
```

I do realize it's kind of confusing to have the class be the same name, let me think about how to rename it.

anonrig commented on packages/core/src/metrics/simpleaggregator.ts:
```diff
@@ -0,0 +1,92 @@
+      return;
+    }
+    if (this._client.captureAggregateMetrics) {
+      const metricBuckets = Array.from(this._buckets).map(([, bucketItem]) => bucketItem);
```
Does Object.values work in this usage?

AbhiPrasad replied:
Nope because ie11 doesn't support it 😢

anonrig commented on packages/core/src/metrics/simpleaggregator.ts:
```diff
@@ -0,0 +1,92 @@
+   */
+  public close(): void {
+    clearInterval(this._interval);
+    this.flush();
```
Flush already clears the bucket. No need for the next line.

anonrig commented on packages/core/src/metrics/simpleaggregator.ts:
```diff
@@ -0,0 +1,92 @@
+function sanitizeTags(unsanitizedTags: Record<string, Primitive>): Record<string, string> {
+  const tags: Record<string, string> = {};
+  for (const key in unsanitizedTags) {
+    if (Object.prototype.hasOwnProperty.call(unsanitizedTags, key)) {
```
Why do we need this check in here? And why not call Object.hasOwn?

anonrig commented on packages/core/src/metrics/utils.ts:
```diff
@@ -0,0 +1,57 @@
+  unit: MeasurementUnit,
+  tags: Record<string, string>,
+): string {
+  const stringifiedTags = Object.entries(dropUndefinedKeys(tags)).sort() as unknown as string[];
```
How does Object.entries().sort() work? Entries return an array of arrays?

AbhiPrasad replied:
It doesn't, will fix.

anonrig commented on packages/core/src/metrics/utils.ts:
```diff
@@ -0,0 +1,57 @@
+  tags: Record<string, string>,
+): string {
+  const stringifiedTags = Object.entries(dropUndefinedKeys(tags)).sort() as unknown as string[];
+  return [metricType, name, unit].concat(stringifiedTags).join('');
```
This creates an intermediary array. We don't need it.

AbhiPrasad replied:
you're right, will fix

anonrig commented on packages/core/src/metrics/utils.ts:
```diff
@@ -0,0 +1,57 @@
+/**
+ * Simple hash function for strings.
+ */
+export function simpleHash(s: string): number {
```
Love this function

AbhiPrasad commented:
Based on Yagiz's PR review I refactored the tag serialization logic and also did a rename of the metrics class.

```ts
Sentry.init({
  dsn: '__DSN__',
  integrations: [
    new Sentry.metrics.MetricsAggregator(),
  ],
});
```

Although the duplication of `metrics` is a little annoying, the namespacing makes intent very clear.
